Adding support for signature verification and public key recovery

brickpop opened the pull request:
Please, review the changes of current Pull Request.

What it does:
- Providing `ecRecover()` and `isValidSignature()`
- Adding unit tests for `ecRecover()` and `isValidSignature()`
- Making `_padTo32()` reusable for other components, so that it can be reused when converting the R/S values of a signature to a `Uint8List`
- `pub run test` runs clean

Please, let me know if you'd like an example added to the Readme or in the `example` folder.

Thank you

brickpop commented:
Is there a reason for all tests to run clean, but the exit code be `1` on Travis?

```
08:24 +162: All tests passed!                                                  
The command "tool/build.sh" exited with 1.

Done. Your build exited with 1.
```

simolus3 commented:
> Is there a reason for all tests to run clean, while Travis completes with an exit code of `1`?

It looks like `lib/src/crypto/secp256k1.dart` needs formatting. Running `dartfmt -w lib/src/crypto/secp256k1.dart` should fix this.

simolus3 commented on lib/src/crypto/secp256k1.dart:
```diff
@@ -95,6 +95,42 @@ MsgSignature sign(Uint8List messageHash, Uint8List privateKey) {
+/// Given an arbitrary message hash and an Ethereum message signature encoded in bytes, returns
+/// the public key that was used to sign it.
+Uint8List ecRecover(Uint8List messageHash, MsgSignature signatureData) {
+  assert(signatureData.r != null);
```
I think we should a comment with a link to https://github.com/web3j/web3j/blob/c0b7b9c2769a466215d416696021aa75127c2ff1/crypto/src/main/java/org/web3j/crypto/Sign.java#L241 as a reference implementation

simolus3 commented on test/crypto/secp256k1_test.dart:
```diff
@@ -46,4 +48,77 @@ void main() {
+        final signature = sign(messageHash, hexToBytes(privateKey));
+
+        expect(
+            isValidSignature(messageHash, signature, originalPublicKey), true,
```
Small nit: Use `isTrue` instead of true (and `isFalse` below).

simolus3 commented on test/crypto/secp256k1_test.dart:
```diff
@@ -46,4 +48,77 @@ void main() {
+              isValidSignature(
+                  messageHash, signature, hexToBytes(invalidPublicKey)),
+              false,
+              reason: 'The signature should be invalid');
```
I would probably add a trailing comma here for nicer formatting with `dartfmt`, but I don't have strong opinions on this.

brickpop commented:
Hi @simolus3, I have addressed all your suggestions. If you are OK with them, It would be so nice to see them making it to pub.dev :)

Thank you for your work!

simolus3 commented:
Thanks for your contribution. I've published this as version `2.0.0-dev.1` on pub.
